aws/resource_aws_cognito_user_pool_client.go

```diff
@@ -0,0 +1,215 @@
+package aws
+
+import (
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsCognitoUserPoolClient() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCognitoUserPoolClientCreate,
+		Read:   resourceAwsCognitoUserPoolClientRead,
+		Update: resourceAwsCognitoUserPoolClientUpdate,
+		Delete: resourceAwsCognitoUserPoolClientDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"generate_secret": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"user_pool_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"explicit_auth_flows": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: false,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validateCognitoUserPoolClientAuthFlows,
+				},
+			},
+
+			"read_attributes": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: false,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"write_attributes": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: false,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"refresh_token_validity": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ForceNew:     false,
+				ValidateFunc: validateIntegerInRange(0, 3650),
+			},
+		},
+	}
+}
+
+func resourceAwsCognitoUserPoolClientCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cognitoidpconn
+
+	params := &cognitoidentityprovider.CreateUserPoolClientInput{
+		ClientName: aws.String(d.Get("name").(string)),
+		UserPoolId: aws.String(d.Get("user_pool_id").(string)),
+	}
+
+	if v, ok := d.GetOk("generate_secret"); ok {
+		params.GenerateSecret = aws.Bool(v.(bool))
+	}
+
+	if v, ok := d.GetOk("explicit_auth_flows"); ok {
+		params.ExplicitAuthFlows = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("read_attributes"); ok {
+		params.ReadAttributes = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("write_attributes"); ok {
+		params.WriteAttributes = expandStringList(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("refresh_token_validity"); ok {
+		params.RefreshTokenValidity = aws.Int64(v.(int64))
```

Comment:
We get `int`, not `int64` from the ResourceData, so this would cause a crash due to invalid cast. I think we'll need to change it to something like this:

```go
params.RefreshTokenValidity = aws.Int64(int64(v.(int)))
```

Reply:
I hadn't actually tested this in AWS and was just going on what VS Code was telling me so I'll take your call.